chirp/drivers/hf90.py

```diff
@@ -0,0 +1,572 @@
+
+        n = _mem.vnum & 0xff
+        if n == 0:
+            n = number
+
+        rsg = RadioSettingGroup("Extra", "extra")
+        rs = RadioSetting("vnum", "Number",
+                          RadioSettingValueInteger(self._lower,
+                                                   self._upper, n))
+        rsg.append(rs)
+        rs = RadioSetting("selcall", "Selcall",
+                          RadioSettingValueBoolean(_mem.selcall))
+        rsg.append(rs)
+        rs = RadioSetting("scan", "Scan",
+                          RadioSettingValueBoolean(_mem.scan))
+        rsg.append(rs)
+        mem.extra = rsg
+
+        return mem
+
+    def _get_tx_freq(self, mem):
+        if mem.duplex == "":
+            return mem.freq
+        elif mem.duplex == "split":
+            return mem.offset
+        elif mem.duplex == "+":
+            return mem.freq + mem.offset
+        elif mem.duplex == "-":
+            return mem.freq - mem.offset
+        return 0
+
+    def set_memory(self, mem):
+        _mem = self._memobj.memory[mem.number - 1]
+
+        _mem.set_raw("\x00" * 8)
+
+        if mem.empty:
+            return
+
+        rx = int(mem.freq / 100)
+        tx = int(self._get_tx_freq(mem) / 100)
+
+        _mem.tx_equals_rx = tx == rx
+
+        _mem.rx0_15 = (rx >> 0) & 0xffff
+        _mem.rx16_19 = (rx >> 16) & 0xf
+
+        _mem.tx0_3 = (tx >> 0) & 0xf
+        _mem.tx4_19 = (tx >> 4) & 0xffff
+
+        _mem.usb = mem.mode == "USB"
+
+        _mem.vnum = mem.number
+
+        for setting in mem.extra:
+            setattr(_mem, setting.get_name(), int(setting.value))
+
+    def validate_memory(self, mem):
+        msgs = chirp_common.CloneModeRadio.validate_memory(self, mem)
+
+        if 0 < self._get_tx_freq(mem) < self._min_tx_freq:
+            msgs.append(chirp_common.ValidationWarning(
+                "tx frequency too low"))
+
+        if mem.extra and isinstance(self, EarlyHF90Radio):
+            if self._memobj.settings.selcall_id == 0x0fff:
+                if mem.extra["scan"].value or mem.extra["selcall"].value:
+                    msgs.append(chirp_common.ValidationWarning(
+                        "standard type but scan or selcall enabled"))
+            else:
+                if mem.extra["scan"].value and not mem.extra["selcall"].value:
+                    msgs.append(chirp_common.ValidationWarning(
+                        "scan enabled but selcall disabled"))
+
+        return msgs
+
+
+# EarlyHF90Radio imitates version 2 of the dealer software whereas
+# LateHF90Radio imitates version 3. It seems both are needed because
+# version 3 warns that it is incompatible with firmware revisions below V301.
+# Unfortunately, the author does not have access to the technical information
+# required to better accomodate the quirks of the various firmware revisions.
+
+@directory.register
+class EarlyHF90Radio(HF90StyleRadio):
+    """Base class for early HF-90 radios"""
+    MODEL = "HF-90 v300 or earlier"
```

Comment:
I haven't looked to see if this changed since I looked last, but.. could I convince you to shorten this? We have a couple of bad examples of super long model strings which have caused various issues in the past with rendering, file naming, etc. Could we make this "HF-90 Early" and "HF-90 Late" and put the version numbers in the info prompt? The info prompt is shown as soon as you select the model, so that should be a reasonable time to explain to the user which is which.

Are there more versions of this radio coming out? I was thinking you said this was effectively dead now... perhaps even "HF-90" and "HF-90 Old" ? TBH I really wish we could eliminate spaces from the model strings entirely just so we know we can consider it a non-breaking short string, but the cat is pretty far out of the bag there.

Reply:
There are no more versions of this radio coming out. What may change in the future is that technical information may come to light which would provide better criteria to distinguish variants and handle their quirks.

It is possible that we will need to add entries to MODEL_COMPAT at some point. But if that's the price of solid technical information (so far unobtainable) it's cheap, IMHO.

I had trouble finding better model names but they can probably be shortened somehow. What's the character count limit that you'd like to impose?

I would not choose "Late" and "Early" for cosmetic reasons unless those were terms already understood by HF-90 owners. Given that late/early is terminology that we invented, it seems inconsiderate to force users to go and look up what we mean by the new terminology.

Whereas, to distinguish models by firmware revision is user-friendly because the firmware revision level is displayed prominently every time the radio powers up (please see my qrz.com page for photos).

Reply:
I just want it to be "shortish", I don't have an actual character limit, but when developing the new UI, it was definitely annoying to get all the models to look right on all three platforms. I also kinda want it to be non-breaking (i.e. no spaces) although as I mentioned, that's somewhat of a moot point.

So, we have another attribute called `VARIANT` that was supposed to be for things like "the american version" or "the european version" of a given radio, where it was basically the same implementation, but with different band limits (or something). That lets us make a distinction between the actual "model" of a thing and some qualifying string/text that distinguishes, well, a variant, which means we could make an informed decision when rendering if needed. In the clone box, it just gets appended to the model string, so it would look the same to the users, but we'd have the opportunity to (potentially) only display the actual "HF-90" in the bottom right status bar, on the website listing, etc.

We don't currently distinguish between those when loading files based on the metadata blob, but I think that should change. Part of the reason I didn't do that originally is because it hasn't always been uniformly applied, but it seems like this is a good example of how it would be used. Maybe I could cook up the change in the directory to make sure it works and then we could consider putting the firmware part in the variant field?

Reply:
As it happens, the first cut of this driver had one model with two variants. I had to abandon that approach as it didn't work correctly. I don't recall exactly why. But I can try again if there's some benfit. If the only benefit is a shorter model name then I guess I'd like to know how short is short enough, as it may still be more easier to have two models with shorter names and no variants.
I agree we'd have to use metadata to determine which of the two variants was applicable to a given image file (the format is ambiguous meaning that it's easy to create an image file that's valid for both variants).
Regarding model names, strictly speaking there's no HF-90. There's HF-90A, HF-90E, HF-90H and HF-90M but it seems best to lump them all together for now (mainly because the vendor PC software is compatible with several of these). Have you given any thought to a compat map for variant strings (similar to MODEL_COMPAT)?

Reply:
> As it happens, the first cut of this driver had one model with two variants. I had to abandon that approach as it didn't work correctly. I don't recall exactly why. But I can try again if there's some benfit. If the only benefit is a shorter model name then I guess I'd like to know how short is short enough, as it may still be more easier to have two models with shorter names and no variants.

To be clear, it's the same number of classes either way. It doesn't really look any different for you from what you have here, except that the firmware part of the string is in `VARIANT` instead of `MODEL`. The only thing I can think of that would have been a problem would have been that loading the image file wouldn't select the proper class.

> I agree we'd have to use metadata to determine which of the two variants was applicable to a given image file (the format is ambiguous meaning that it's easy to create an image file that's valid for both variants). Regarding model names, strictly speaking there's no HF-90. There's HF-90A, HF-90E, HF-90H and HF-90M but it seems best to lump them all together for now (mainly because the vendor PC software is compatible with several of these).

Yeah, the reason it matters is so that when someone selects a model/variant (i.e. a class) that we keep that consistent through the download, save, open, upload chain. Not knowing the differences in the A and E models (for example) I can't really comment, but the only reason to distinguish would be when things need to be handled differently (i.e. the cloning or specific rules). If you can determine what firmware version the radio is from the cloning process and/or the image file, I'd say we need only one class and you can just handle the behavioral differences internally. I assume that's the case for the letter differences but for the firmware differences you need the distinct implementations, right? If so, model/variant (and thus separate classes) are the right way.

> Have you given any thought to a compat map for variant strings (similar to MODEL_COMPAT)?

No because it hasn't mattered until now. `MODEL_COMPAT` is just there because we had to change the model name of a few radios and we needed to honor the old format when loading so we didn't make old image files un-openable after the change. We had successfully avoided that for like 14 years until that point, and we definitely should avoid it if at all possible (which is why I want to get this model thing sorted before merge). But, if we *had* to make a variant change, then yes, we'd need to extend that mapping to be a 3-tuple of vendor,model,variant.

Reply:
I just noticed your question about detecting a radio's firmware revision level. I don't think it is feasible. It is possible to get the radio to dump any 16 byte region in the address range 0x0000 through 0x1fff but I believe that this doesn't cover all the relevant memory, and figuring out which bytes can be used to fingerprint the radio (model and firmware rev) is not feasible without access to either technical  documentation or a large variety  radios to examine. The bottom line is that the CHIRP user must choose.
Regarding your `directory-variant` branch, I rebased this branch onto that one and the `VARIANT` conversion was straight-forward.
My only concern would be old img files out there, created by old versions of the existing drivers. The trick would be to ensure that old metadata works with the current drivers, in cases where the VARIANT strings for a given driver have changed over time...